accelerator/core/serializer/serializer.c

```diff
@@ -696,6 +718,101 @@ status_t ta_find_transactions_by_tag_res_serialize(const ta_find_transactions_by
   return ret;
 }
 
+static status_t send_mam_message_mam_v1_req_deserialize(cJSON const* const json_obj, ta_send_mam_req_t* const req) {
+  cJSON *json_key = NULL, *json_value = NULL;
+  status_t ret = SC_OK;
+  send_mam_data_mam_v1_t* data = (send_mam_data_mam_v1_t*)req->data;
+  send_mam_key_mam_v1_t* key = (send_mam_key_mam_v1_t*)req->key;
+
+  ret = ta_json_get_string(json_obj, "x-api-key", req->service_token, SERVICE_TOKEN_LEN);
+  if (ret != SC_OK) {
+    ta_log_error("%d\n", ret);
+    goto done;
+  }
+
+  if (cJSON_HasObjectItem(json_obj, "key")) {
+    json_key = cJSON_GetObjectItem(json_obj, "key");
+    if (json_key == NULL) {
+      ret = SC_SERIALIZER_JSON_PARSE;
+      ta_log_error("%s\n", "SC_SERIALIZER_JSON_PARSE");
+      goto done;
+    }
+
+    if (cJSON_HasObjectItem(json_key, "psk")) {
+      ret = json_string_array_to_string_utarray(json_key, "psk", key->psk_array);
+      if (ret != SC_OK) {
+        ret = SC_SERIALIZER_JSON_PARSE;
+        ta_log_error("%s\n", "SC_SERIALIZER_JSON_PARSE");
+        goto done;
+      }
+    }
+
+    if (cJSON_HasObjectItem(json_key, "ntru")) {
+      ret = json_string_array_to_string_utarray(json_key, "ntru", key->ntru_array);
+      if (ret != SC_OK) {
+        ret = SC_SERIALIZER_JSON_PARSE;
+        ta_log_error("%s\n", "SC_SERIALIZER_JSON_PARSE");
+        goto done;
+      }
+    }
+  }
+
+  json_key = cJSON_GetObjectItem(json_obj, "data");
+  if (json_key == NULL) {
+    ret = SC_SERIALIZER_JSON_PARSE;
+    ta_log_error("%s\n", "SC_SERIALIZER_JSON_PARSE");
+    goto done;
+  }
+
+  json_value = cJSON_GetObjectItemCaseSensitive(json_key, "seed");
+  if (json_value != NULL) {
+    size_t seed_size = strnlen(json_value->valuestring, NUM_TRYTES_ADDRESS);
+
+    if (seed_size != NUM_TRYTES_HASH) {
+      ret = SC_SERIALIZER_INVALID_REQ;
+      ta_log_error("%s\n", "SC_SERIALIZER_INVALID_REQ");
+      goto done;
+    }
+    data->seed = (tryte_t*)malloc(sizeof(tryte_t) * (NUM_TRYTES_ADDRESS + 1));
+    snprintf((char*)data->seed, seed_size + 1, "%s", json_value->valuestring);
+  }
+
+  json_value = cJSON_GetObjectItemCaseSensitive(json_key, "message");
+  if (json_value != NULL) {
+    size_t payload_size = strlen(json_value->valuestring);
+    char* payload = (char*)malloc((payload_size + 1) * sizeof(char));
+
+    // In case the payload is unicode, the character whose ASCII code is beyond 128 result to an
+    // error status_t code
+    for (size_t i = 0; i < payload_size; i++) {
+      if (json_value->valuestring[i] & 0x80) {
+        ret = SC_SERIALIZER_JSON_PARSE_ASCII;
+        ta_log_error("%s\n", "SC_SERIALIZER_JSON_PARSE_ASCII");
+        goto done;
```

Comment:
Can you simply return?

Reply:
No we can't. The cJSON objects need to be freed.